### ISSUE TITLE
Better compatibility with batch normalization

### ISSUE THREAD

**Issue by andreped** [andreped/GradientAccumulator#47]
As you all know, gradient accumulation is not directly compatible with batch normalization, as batch normalization will be updated for every single forward step and we cannot control that externally.

In order to get the same behaviour as for gradient updates, we will likely need to implement a custom batch normalization layer which does this internally, as overloading the original batch norm step seems challenging (due to its extreme complexity).

**Comment by andreped** [comment]
I implemented a custom batch norm layer recently, which was made available in [v0.3.2](https://github.com/andreped/GradientAccumulator/releases/tag/v0.3.2).

Implementation can be seen [here](https://github.com/andreped/GradientAccumulator/blob/c42100db510a401a7f09c9a6f1293953216fef0f/gradient_accumulator/layers.py#L9).

The plan is to add accumulation to the `call` step similar as done in the Model wrapper.

The performance on the custom layer reaches very similar results to keras' BN layer, but we should get identical before we can say that it is working as intended.

**Comment by andreped** [comment]
To add accumulation support to the custom BN layer, relevant resources are [this](https://github.com/keras-team/keras/blob/v2.12.0/keras/layers/normalization/batch_normalization.py#L586), [this](https://www.tensorflow.org/api_docs/python/tf/nn/batch_normalization), and [this](https://github.com/tensorflow/tensorflow/blob/v2.12.0/tensorflow/python/ops/nn_impl.py#L1528-L1596).

**Comment by andreped** [comment]
I have added gradient accumulation support to the custom BN layer now in 05fb499078408283c4544041b166c8d8963cd1ad.

However, unit tests show that the results are not equivalent when increasing `accum_steps`. Hence, it is not yet working as intended.

**Comment by andreped** [comment]
Support for gradient accumulation to batch normalization has been added. Even though it is not perfect, at least people could test it and improve it further.

I also updated the documentations regarding how to use it in f54e389d1d2979cbf0409ddab4a726e58367fcae.

**Comment by axeldavy** [comment]
Hi,

When looking at the proposed batch normalization, I get the impression the main modification is that the running average and variance are updated less frequently. As these are only used at test time, this should not impact training. Besides a similar effect can be had by changing the multiplication factor in the running average to average on more batches.

Maybe I'm missing something of course.

One suggestion would be to implement a batch normalization known to work well with small batches.
For example Yolov4 (https://arxiv.org/pdf/2004.10934.pdf), a well known object detector, implements Cross mini-Batch Normalization for this purpose.

**Comment by axeldavy** [comment]
In my understanding of the code, the running mean and variance are the only parts that are updated less frequently in this version. And these are only used at test time. The batch mean and var is still used for the forward pass, as a normal BatchNormalization.

**Comment by andreped** [comment]
The modifications I did to the BN layer, are only relevant for training. It should work as expected for inference, regardless of gradient accumulation. But then again, I have yet to properly benchmark it. So little time on my plate for open source recently sadly
